models/curvefi/optimism/curvefi_optimism_trades.sql

```diff
@@ -0,0 +1,202 @@
+            substring(topic2,25,40) AS taker,
+            '' AS maker,
+            conv(substring(data,3+64*3,64),16,10) as token_bought_amount_raw, --2nd bought
+            conv(substring(data,3+64*1,64),16,10) as token_sold_amount_raw, --1st sold
+            contract_address AS project_contract_address,
+            l.tx_hash,
+            '' AS trace_address,
+            index AS evt_index,
+            conv(substring(data,3+64*2,64),16,10) AS bought_id,
+            conv(substring(data,3+64*0,64),16,10) AS sold_id
+        FROM {{ source('optimism', 'logs') }} l
+        INNER JOIN {{ ref('curvefi_optimism_pools') }} ta
+            ON l.contract_address = ta.pool
+            AND conv(substring(data,3+64*3,64),16,10) = ta.tokenid --t.bought_id = ta.tokenid
+            AND ta.version = 'Basic Pool'
+        INNER JOIN {{ ref('curvefi_optimism_pools') }} tb
+            ON l.contract_address = tb.pool
+            AND conv(substring(data,3+64*1,64),16,10) = tb.tokenid --t.sold_id = tb.tokenid
+            AND tb.version = 'Basic Pool'
+        WHERE l.topic1 = '0x8b3e96f2b889fa771c53c981b40daf005f63f637f1869f707052d15a3dd97140'
+        AND l.topic2 IS NOT NULL AND l.data IS NOT NULL
+        {% if is_incremental() %}
+        AND l.block_time >= date_trunc('day', now() - interval '1 week')
+        {% endif %}
+        {% if not is_incremental() %}
+        AND l.block_time >= '{{project_start_date}}'
+        {% endif %}
+    ) cp
+    INNER JOIN {{ ref('curvefi_optimism_pools') }} ta
+        ON cp.project_contract_address = ta.pool
+        AND cp.bought_id = ta.tokenid
+    INNER JOIN {{ ref('curvefi_optimism_pools') }} tb
+        ON cp.project_contract_address = tb.pool
+        AND cp.sold_id = tb.tokenid
+    LEFT JOIN {{ ref('tokens_optimism_erc20') }} ea
+        ON ea.contract_address = ta.token
+    LEFT JOIN {{ ref('tokens_optimism_erc20') }} eb
+        ON eb.contract_address = tb.token
+)
+SELECT DISTINCT
+    'optimism' AS blockchain,
+    'curve' AS project,
+    '1' AS version,
+    TRY_CAST(date_trunc('DAY', dexs.block_time) AS date) AS block_date,
+    dexs.block_time,
+    COALESCE(erc20a.symbol,p_bought.symbol)AS token_bought_symbol,
+    COALESCE(erc20b.symbol,p_sold.symbol) AS token_sold_symbol
+    ,case
+        when lower(COALESCE(erc20a.symbol,p_bought.symbol)) > lower(COALESCE(erc20b.symbol,p_sold.symbol)) then concat(COALESCE(erc20b.symbol,p_sold.symbol), '-', COALESCE(erc20a.symbol,p_bought.symbol))
+        else concat(COALESCE(erc20a.symbol,p_bought.symbol), '-', COALESCE(erc20b.symbol,p_sold.symbol))
+    end as token_pair,
+    --metapools seem to always use the added coin's decimals if it's the one that's bought - even if the other token has less decimals (i.e. USDC)
+    dexs.token_bought_amount_raw / POWER(10 , (CASE WHEN pool_type = 'meta' AND bought_id = 0 THEN COALESCE(erc20b.decimals,p_sold.decimals) ELSE COALESCE(erc20a.decimals,p_bought.decimals) END) ) AS token_bought_amount,
```

Comment:
gm @MSilb7 !

I am looking into an inconsistency we are having on this spell, [where some trades are overstated](https://dune.com/queries/1771927?d=1). It seems the issue boils down to this line, where the wrong number of decimals are being applied.

Do you know why this logic exists? I don't see it in the ethereum spell.

Reply:
hey, yeah on metapools, it seemed like the decimals in the number emitted mapped to the token bought rather than each token. Maybe worth looking at a few trades again to make sure this didn't flip back or something?

Reply:
Event - Tokens look like they have the same decimals, even though USDC is 6 and sUSD is 18

<img width="1356" alt="image" src="https://user-images.githubusercontent.com/4006780/208255502-96bd0cdb-c0c5-4c8b-8c05-987541c7b9af.png">

But the event seems misconfigured? 3crv is being swapped, not usdc
<img width="845" alt="image" src="https://user-images.githubusercontent.com/4006780/208255579-f5a785e1-4ee5-400a-ba84-d7428a60788a.png">


Maybe the issue is that it will always use the curve pool's amount of decimals (18) rather than the underlying swapped token amount of decimals (6). Will test...

Reply:
Yeah, confirmed the meta swaps just use the curve pool's decimals:
https://dune.com/queries/1772295

Will update to make it "18 if bought_id = 0"

Reply:
PR with the https://github.com/duneanalytics/spellbook/pull/2345 @augustog